File: deck/base/code/native/browser/runtime/crypto.ts
const crypto = (() => {
  // digest / hmac take and return raw bytes (Uint8Array), the crypto currency; random is hex; hex is an edge codec
  const hex = (buffer) => Array.from(new Uint8Array(buffer)).map((b) => b.toString(16).padStart(2, '0')).join('')
  const digest = async (algorithm, input) => new Uint8Array(await globalThis.crypto.subtle.digest(algorithm, input))
  const mac = async (algorithm, key, data) => {
    const cryptoKey = await globalThis.crypto.subtle.importKey('raw', key, { name: 'HMAC', hash: algorithm }, false, ['sign'])
    return new Uint8Array(await globalThis.crypto.subtle.sign('HMAC', cryptoKey, data))
  }
  return {
    sha256: (input) => digest('SHA-256', input),
    sha512: (input) => digest('SHA-512', input),
    md5: (input) => { throw new Error('MD5 is not available in the Web Crypto API') },
    hmacSha256: (key, data) => mac('SHA-256', key, data),
    hmacSha512: (key, data) => mac('SHA-512', key, data),
    randomBytes: (size) => hex(globalThis.crypto.getRandomValues(new Uint8Array(size)).buffer),
  }
})()

File: deck/seed/code/native/rust/runtime/crypto.rs
mod crypto {
    use sha2::{Sha256, Sha512, Digest};
    use hmac::{Hmac, Mac};
    pub fn sha256(input: Vec<u8>) -> Vec<u8> { Sha256::digest(&input).to_vec() }
    pub fn sha512(input: Vec<u8>) -> Vec<u8> { Sha512::digest(&input).to_vec() }
    pub fn md5(input: Vec<u8>) -> Vec<u8> { ::md5::Md5::digest(&input).to_vec() }
    pub fn hmac_sha256(key: Vec<u8>, data: Vec<u8>) -> Vec<u8> {
        let mut mac = Hmac::<Sha256>::new_from_slice(&key).unwrap();
        mac.update(&data);
        mac.finalize().into_bytes().to_vec()
    }
    pub fn hmac_sha512(key: Vec<u8>, data: Vec<u8>) -> Vec<u8> {
        let mut mac = Hmac::<Sha512>::new_from_slice(&key).unwrap();
        mac.update(&data);
        mac.finalize().into_bytes().to_vec()
    }
    pub fn random_bytes(size: i64) -> Vec<u8> {
        use ::rand::RngCore;
        use ::rand::rngs::OsRng;
        let mut buffer = vec![0u8; size as usize];
        OsRng.fill_bytes(&mut buffer);
        buffer
    }
}
